### ISSUE TITLE
Phase4 Deduplication broken?

### ISSUE THREAD

**Issue by tfmorris** [dkpro/dkpro-c4corpus#41]
I'm only going by inspection here, but I don't think this actually works.

The mappers take two input streams and tag them with their source (0=WARC, 1=dupe list) then output to the reducer using a composite key consisting of the WARC ID plus the source.

The reducer looks at how many values a key has and only outputs the document if the number of values is 1. This would work if the key was just the WARC ID, but since it includes the source tag as well, it's actually going to add the dummy documents to the output stream rather than deleting the duplicate documents.

I've reorganized the data flow and collapsed phases 1, 2, 3.1, 3.2, 3.3, 3.4 into a single job which generates both text-only WARCs and a list of duplicate document IDs and was converted Phase 4 into my new phase 2 when I ran across this.

If you're open to adopting my new workflow (when it's ready to be reviewed), this can probably be deferred, but if you're going to use the existing pipeline for a while, someone should have a look and see if I'm confused or this is a real problem.

**Comment by habernal** [comment]
There are two sources (aka two mappers) = one is the full warc collection from previous steps (with possible duplicate records) and one is a txt list of record IDs that were detected as duplicate in Phase3Step4. Both mappers emit warc record (either "normal" warc or "dummy" warc) as value using the same composite key (warc record ID). So if there are two items with the same warc record ID in the reducer, this means that this particular warc record is duplicate and is omitted.

Phases 1, 2, and 3 can be merged into one big M/R job; we kept them separately to debug issues. The new implementation can also be kept parallel to the existing ones; feel free to open a new separate issue for that.

**Comment by tfmorris** [comment]
I missed the SourceJoiningGroupingComparator which basically ignores everything that the CompositeKey adds. My bad! The existing implementation should work fine.

You could get the same effect, without the extra complexity, by dropping both CompositeKey and it's custom comparator and just using a Text key consisting of the WARC ID.

In my merged phase 4 & 5, I actually add the language/license/etc to the text key so that the sorting and deduping can be done together, cutting the 8 separate jobs down to 2. I'll open a PR in parallel with my final testing so that folks can start reviewing it.
